src/function/aggregate/holistic/quantile.cpp

```diff
@@ -211,73 +266,85 @@ struct DiscreteQuantileListOperation : public QuantileOperation<INPUT_TYPE> {
 		for (const auto &quantile : bind_data->quantiles) {
 			auto offset = (idx_t)((double)(state->pos - 1) * quantile);
 			std::nth_element(v_t, v_t + offset, v_t + state->pos);
-			auto val = Value::CreateValue(v_t[offset]);
+			auto val = CAST::CreateValue(v_t[offset]);
 			ListVector::PushBack(result_list, val);
 		}
 		target[idx].length = bind_data->quantiles.size();
 	}
 };
 
-AggregateFunction GetDiscreteQuantileListAggregateFunction(PhysicalType type) {
-	switch (type) {
-	case PhysicalType::INT16:
-		return QuantileListAggregate<QuantileState, int16_t, list_entry_t, DiscreteQuantileListOperation<int16_t>>(
-		    LogicalType::SMALLINT, LogicalType::SMALLINT);
-
-	case PhysicalType::INT32:
-		return QuantileListAggregate<QuantileState, int32_t, list_entry_t, DiscreteQuantileListOperation<int32_t>>(
-		    LogicalType::INTEGER, LogicalType::INTEGER);
-
-	case PhysicalType::INT64:
-		return QuantileListAggregate<QuantileState, int64_t, list_entry_t, DiscreteQuantileListOperation<int64_t>>(
-		    LogicalType::BIGINT, LogicalType::BIGINT);
+template <typename INPUT_TYPE, class CAST = Value>
+AggregateFunction GetTypedDiscreteQuantileListAggregateFunction(const LogicalType &type) {
+	return QuantileListAggregate<QuantileState, INPUT_TYPE, list_entry_t,
+	                             DiscreteQuantileListOperation<INPUT_TYPE, CAST>>(type, type);
+}
 
-	case PhysicalType::INT128:
-		return QuantileListAggregate<QuantileState, hugeint_t, list_entry_t, DiscreteQuantileListOperation<hugeint_t>>(
-		    LogicalType::HUGEINT, LogicalType::HUGEINT);
-	case PhysicalType::FLOAT:
-		return QuantileListAggregate<QuantileState, float, list_entry_t, DiscreteQuantileListOperation<float>>(
-		    LogicalType::FLOAT, LogicalType::FLOAT);
+AggregateFunction GetDiscreteQuantileListAggregateFunction(const LogicalType &type) {
+	switch (type.id()) {
+	case LogicalTypeId::TINYINT:
```

Comment:
Same here, this could be using the physical type only no?

Reply:
Same answer ;-)

Reply:
Actually, no. Soon we are going to have `struct` types and this will not work for `Value::CreateValue`. I think we need to rely on the kinds of the linker...